### ISSUE TITLE
Incorrect use of DOES>

### ISSUE THREAD

**Issue by ruv** [DouglasBHoffman/FMS2#1]
https://github.com/DouglasBHoffman/FMS2/blob/4fe1b82e3855ae2f29f5ea489b5522bfbdfb88f7/FMS2LL/FMS2LL.f#L216-L219

If [`does>`](https://forth-standard.org/standard/core/DOES) is used inside control structures, then the expected parameter `colon-sys1` is not available on the top of the control-flow stack according to the stack diagram `( C: colon-sys1 -- colon-sys2 )` for its compilation semantics.

A possible correction:
```forth
  exit then
  \ outside a class definition, so instantiate a new named dictionary object
  create immediate (dict)-xt execute drop
  does> state @ if ?execute then ;
```

**Comment by DouglasBHoffman** [comment]
Outstanding!  The problem I was having with Gforth was self inflicted. Thank you for taking a look and catching my error.

As you can see, my design decision regarding late binding and early binding is to default to late binding EXCEPT in the cases where the FMS compiler knows the class of the object at compile time. This can result in a significant speed increase with no extra effort required by the programmer. The use of ESELF instead of SELF is the exception where the programmer specifies inside a method definition. In some cases where it is clear that a subclass will not benefit from a late bind to SELF I think ESELF is justified and useful.

The other thing I don’t provide in FMS2 is message-less instance variable access. No one has asked for it. It could be done if someone wanted it. Temporarily changing the wordlist search order would be required.


If it is ok with you I would like to incorporate the correction you provide below and attribute that to you in the source.

Thanks again!

-Doug

**Comment by ruv** [comment]
> If it is ok with you I would like to incorporate the correction you provide below and attribute that to you in the source.

Of course, it's OK. Otherwise, why do I provide the code? -)

Concerning attributions — it's up to you. But I think, no need for any attribution in the source, it's enough to place it in the commit message (if any). Rationale: if some source code will contain many contributions, it will contain too many attributions, one over another.

**Comment by ruv** [comment]
Closed by commit 354b55c800f1800
